File: compute/backend/objects/rule.py
from dataclasses import dataclass

@dataclass
class Rule:
    """
    Represents a scheduling rule for when a habit should occur or be checked.

    A rule specifies a day of the week and a time (hour + minute), along with
    whether the rule is currently active. These rules can be used for reminders,
    automations, or inferred behavior based on historical habit logs.

    Attributes
    ----------
    day : int
        Day of the week (typically 0-6, where 0 = Monday or Sunday depending on convention).
    hour : int
        Hour of the day in 24-hour format (0-23).
    minute : int
        Minute of the hour (0-59).
    active : bool
        Whether this rule is enabled and should be applied.
    """
    day: int
    hour: int
    minute: int
    active: bool

File: compute/mobile/habittrackerapp/lib/rule.dart
class Rule {
  final int id;
  final String habit;
  final int day;
  final int hour;
  final int minute;
  final bool active;

  Rule({
    required this.id,
    required this.habit,
    required this.day,
    required this.hour,
    required this.minute,
    required this.active,
  });

  factory Rule.fromJson(Map<String, dynamic> json) {
    return Rule(
      id: json['id'],
      habit: json['habit'],
      day: json['day'],
      hour: json['hour'],
      minute: json['minute'],
      active: json['active'],
    );
  }

  Map<String, String> toQuery() {
    return {
      'habit': habit,
      'day': day.toString(),
      'hour': hour.toString(),
      'minute': minute.toString(),
      'active': active.toString(),
    };
  }
}
